components/api/src/main/kotlin/com/hotels/styx/api/MeterRegistry.kt

```diff
@@ -0,0 +1,267 @@
+import java.util.*
+import java.util.concurrent.TimeUnit
+import java.util.function.Consumer
+import java.util.function.ToDoubleFunction
+import java.util.function.ToLongFunction
+
+/**
+ * An interface to represent any implementation of a meter registry.
+ */
+interface MeterRegistry {
+    val meters: List<Meter>
+
+    fun counter(name: String, vararg tags: String): Counter
+    fun counter(name: String, tags: Iterable<Tag>): Counter
+    fun summary(name: String, vararg tags: String): DistributionSummary
+    fun summary(name: String, tags: Iterable<Tag>): DistributionSummary
+    fun get(name: String): RequiredSearch
+    fun find(name: String): Search
+    fun forEachMeter(consumer: Consumer<Meter>)
+    fun config(): Config
+    fun timer(name: String, tags: Iterable<Tag>): Timer
+    fun timer(name: String, vararg tags: String): Timer
+    fun more(): More
+    fun <T> gauge(name: String, tags: Iterable<Tag>, stateObject: T, valueFunction: ToDoubleFunction<T>): T
+    fun <T : Number> gauge(name: String, tags: Iterable<Tag>, number: T): T
+    fun <T : Number> gauge(name: String, number: T): T
+    fun <T> gauge(name: String, stateObject: T, valueFunction: ToDoubleFunction<T>): T
+    fun <T : Collection<*>> gaugeCollectionSize(name: String, tags: Iterable<Tag>, collection: T): T
+    fun <T : Map<*, *>> gaugeMapSize(name: String, tags: Iterable<Tag>, map: T): T
+    fun remove(meter: Meter): Meter?
+    fun remove(mappedId: Meter.Id): Meter?
+    fun clear()
+    fun close()
+    fun isClosed(): Boolean
+
+    interface More {
+        fun longTaskTimer(name: String, vararg tags: String): LongTaskTimer
+        fun longTaskTimer(name: String, tags: Iterable<Tag>): LongTaskTimer
+        fun <T> counter(name: String, tags: Iterable<Tag>, obj: T, countFunction: ToDoubleFunction<T>): FunctionCounter
+        fun <T : Number> counter(name: String, tags: Iterable<Tag>, number: T): FunctionCounter
+        fun <T> timer(
+            name: String,
+            tags: Iterable<Tag>,
+            obj: T,
+            countFunction: ToLongFunction<T>,
+            totalTimeFunction: ToDoubleFunction<T>,
+            totalTimeFunctionUnit: TimeUnit
+        ): FunctionTimer
+
+        fun <T> timeGauge(name: String, tags: Iterable<Tag>, obj: T, timeFunctionUnit: TimeUnit, timeFunction: ToDoubleFunction<T>): TimeGauge
+    }
+
+    // Functions below here are not for general metric handling, but needed by Styx internally.
+
+    /**
+     * Creates a new scoped registry based on this registry. If this registry is already scoped, the new scope will be prefixed to the existing scope.
+     *
+     * So, if `fooRegistry` has scope "foo", and we call `val barRegistry = fooRegistry.scope("bar")`, then `barRegistry` will have the scope "bar.foo".
+     */
+    fun scope(scope: String): ScopedMeterRegistry = ScopedMeterRegistry(this, scope)
+    fun startTimer(): Timer.Sample
+    fun micrometerRegistry(): io.micrometer.core.instrument.MeterRegistry?
```

Comment:
I think maybe we should remove the '?' from the line above as the micrometer registry should never be null

Reply:
OK, I've taken a quick look at the rest of the code, which all assumes that the return value is non-null anyway, so I will remove the `?`.